### DISCUSSION TITLE
Migration to JointJS v4: Issue with graph.fromJSON()

### DISCUSSION THREAD

**Discussion by VADDIJASWANT** [clientIO/joint#2566] [Q&A]
### Introduction

Hello everyone,

I am currently migrating to JointJS version 4 and I am facing an issue with the graph.fromJSON() method. In my project, I have nearly 900 SVGs that I usually load to the JointJS paper as follows:
```
const { dia, ui, shapes, util, g } = joint;
shapes.custom = {}
pub.create_svg_element_template = (item, svgString) => {
    joint.shapes.custom[item.name] = joint.dia.Element.define(`${item.name}`, {
        customType:'SVGS',
        attrs: {
            svg: {
            },
            label: {
            },
            additional_properties: {
                
            },

        }
    }, {
        markup: util.svg`${svgString}`,
        toJSON: function () {
            // Call the parent toJSON method to get the default serialization
            const json = joint.dia.Element.prototype.toJSON.apply(this, arguments);
            json.markup = this.markup;
            json.attrs = this.get('attrs');
            return json;
        },
    });

}
// for custom svgs which are uploaded from custom shapes
pub.creating_custom_svg = (item,x,y) => {
    item = JSON.parse(item)
    if (!shapes.custom.hasOwnProperty(item.name)) {
        pub.create_svg_element_template(item, svgString)
    }

    var SVG = new joint.shapes.customitem.name;
    SVG.position(x, y);
    SVG.resize(100, 100);
    graph.addCell(SVG);

}
```
In the database, I am saving the JSON string of the graph. In previous version Even when I restart the project don't load any shapes or svgs to the paper I was able to retrieve the old graph using fromJSON(), it works properly and the graph loads correctly. However, when I use v4.0, I get errors stating dia.Graph: Could not find cell constructor for type: 'TagWidget'. Make sure to add the constructor to 'cellNamespace'.

I would like to know if there is any difference between the fromJSON method from before to now, or if I am missing anything. Any help would be greatly appreciated.

Thank you.

### Does your question relate to JointJS or JointJS+. Select both if applicable.

JointJS, JointJS+

**Comment by VADDIJASWANT** [comment]
Hello everyone for discussion I raised above regarding graph.fromJson in JointJS v4.0. I found that in changelog and according to it, elements retrieved from a JSON string must be predefined in v4.0. like `const MyElement = dia.Element.define('MyElement');` However, I’m facing a small issue with this.

I have a very large number of custom SVGs (approximately 1000) in my library.At the start of the project, as I want to retrieve any particular saved  graph and add it to the paper. As I won’t know which SVGs are being used in that graph I cannot write code only for the required svgs, even if I did define it in the format of `const MyElement = dia.Element.define('MyElement');`
it is ok for the retrieved graph as markup for the element is defined in the json string but when I try to add svg manually to graph I am given a error that markup is required,

so the one way I thought to resolve this issue is to at the start of the project define all the custom elements with the required markup and properties, but as I have nearly 1000 svgs it will effect the performance of the project 

Is there a more efficient way to define all custom elements in dia.Elements at the start of the project? Defining all elements upfront could significantly impact project performance. Any suggestions would be appreciated.

**Reply by kumilingus** [reply]
Could you share an example with us that can be run?

For instance you can have a single Element defined e.g. `MyElement` and only set `markup` attribute.
```js
const MyElement = joint.dia.Element.define('MyElement');
// define markup via `attribute` not on `prototype`.
const el1 = new MyElement({ appType: item1.name, markup: markups[item1.name]  });
const el2 = new MyElement({ appType: item2.name, markup: markups[item2.name]  });
```

**Reply by VADDIJASWANT** [reply]
[JointJSGraphfromJson.zip](https://github.com/clientIO/joint/files/14599955/JointJSGraphfromJson.zip)
Hi @kumilingus,

I have attached a small zip file for you to download. Please add joint-plus.min.js and joint-plus.min.css to the root folder. I did not include them here as they are licensed.

At the end of the script, you will find a small commented code. Once you uncomment it and run it, you will encounter the issue.

I am sure we are able to parse the JSON string, retrieve the object, loop through it, and initialize all the unique custom elements used in the graph. However, lets assume if there are 200 unique custom elements on the graph, I need to loop a minimum of 200 times and initialize them. This could affect performance.

In the past with version 3.7, we were able to retrieve the graph directly without defining the elements, which saved time and improved performance. Now, I am trying to find a solution that maintains this efficiency.

**Reply by M27LAD** [reply]
@VADDIJASWANT - this may or may not help, but i was coming across a similar problem, but then found this page: [https://resources.jointjs.com/tutorial/cell-namespace]

I hadn't added the namespace when defining the graph, so when trying to import JSON I received the Cell Namespace error.
In your case it looks like you may need to add the custom elements to your name space when defining the graph - but as you mention, you have over 200 custom elements, therefore this approach may not be feasible!

**Reply by kumilingus** [reply]
@VADDIJASWANT, your shape is missing SVG markup. And that's exactly what the error says.

For instance, this will make the error go away:
```ts
const TagWidget = dia.Element.define('TagWidget', {}, { markup: util.svg`<rect/>` });
```
